File: product_of_array_except_self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        answer = [1] * n
        for i in range(1,n):
            answer[i] = answer[i-1] * nums[i-1]

        right = 1
        for i in range(n-1, -1, -1):
            answer[i] *= right
            right *= nums[i]

        return answer

File: product_of_array_except_self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {

        int n = nums.size();
        vector<int> answer(n,1);

        for (int i = 1; i < nums.size(); i++){

            answer[i] = answer[i-1] * nums[i-1];
        }
        int right = 1;
        for (int j = nums.size() - 1; j >= 0; j--){

            answer[j] *= right;
            right *= nums[j];

        }

        return answer;

        
    }
};
